include/circt/Dialect/RTGTest/IR/RTGTestOps.td

```diff
@@ -275,8 +278,192 @@ class InstFormatSOpBase<string mnemonic, int opcode7, int funct3>
+              .getRegisterAssembly()
+         << ", ";
+
+      if (auto label = dyn_cast<StringAttr>(adaptor.getImm())) {
+        os << label.getValue();
+        return;      
+      }
+
+      // The assembler wants an unsigned value here.
+      os << cast<Imm32Attr>(adaptor.getImm()).getValue();
+    }
+  }];
+}
+
+class InstFormatJOpBase<string mnemonic, int opcode7>
+  : RTGTestOp<"rv32i." # mnemonic, [InstructionOpAdaptor]> {
+
+  let arguments = (ins IntegerRegisterType:$rd,
+                       AnyTypeOf<[Imm21Type, LabelType]>:$imm);
+
+  let assemblyFormat = "$rd `,` $imm `:` type($imm) attr-dict";
+
+  let extraClassDefinition = [{
+    void $cppClass::printInstructionBinary(llvm::raw_ostream &os,
+                                           FoldAdaptor adaptor) {
+      assert (isa<Imm21Attr>(adaptor.getImm()) &&
+              "binary of labels not supported");
+
+      auto rd = cast<rtg::RegisterAttrInterface>(adaptor.getRd());
+      auto imm = cast<Imm21Attr>(adaptor.getImm()).getAPInt();
+
+      auto binary = imm.extractBits(1, 20)
+      .concat(imm.extractBits(10, 1))
+      .concat(imm.extractBits(1, 1))
+      .concat(imm.extractBits(8, 12))
+      .concat(llvm::APInt(5, rd.getClassIndex()))
+      .concat(llvm::APInt(7, }] # opcode7 # [{));
+
+      SmallVector<char> str;
+      binary.toStringUnsigned(str, 16);
+      os << str;
+    }
+
+    void $cppClass::printInstructionAssembly(llvm::raw_ostream &os,
+                                             FoldAdaptor adaptor) {
+      os << getOperationName().rsplit('.').second
+         << " "
+         << cast<rtg::RegisterAttrInterface>(adaptor.getRd())
+              .getRegisterAssembly()
+         << ", ";
+
+      if (auto label = dyn_cast<StringAttr>(adaptor.getImm())) {
+        os << label.getValue();
+        return;      
+      }
+
+      // The assembler is fine with signed and unsigned values here.
+      os << cast<Imm21Attr>(adaptor.getImm()).getAPInt().getSExtValue();
+    }
+  }];
+}
+
+class InstFormatIAOpBase<string mnemonic, int opcode7, int funct3>
+  : RTGTestOp<"rv32i." # mnemonic, [InstructionOpAdaptor]> {
+
+  let arguments = (ins IntegerRegisterType:$rd,
+                       IntegerRegisterType:$rs,
+                       Imm12Type:$imm);
+
+  let assemblyFormat = "$rd `,` $rs `,` $imm attr-dict";
+
+  let extraClassDefinition = [{
+    void $cppClass::printInstructionBinary(llvm::raw_ostream &os,
+                                           FoldAdaptor adaptor) {
+      auto rd = cast<rtg::RegisterAttrInterface>(adaptor.getRd());
+      auto rs = cast<rtg::RegisterAttrInterface>(adaptor.getRs());
+      auto imm = cast<Imm12Attr>(adaptor.getImm()).getAPInt();
+
+      auto binary = imm
+      .concat(llvm::APInt(5, rs.getClassIndex()))
+      .concat(llvm::APInt(3, }] # funct3 # [{))
+      .concat(llvm::APInt(5, rd.getClassIndex()))
+      .concat(llvm::APInt(7, }] # opcode7 # [{));
```

Comment:
This _might_ be easier to read with a TableGen `subst` call around the code snippet substituting something like `$funct3`. But then again, this is pretty clean, and there are only two values to be interpolated. The substitution might make it harder to read.

Reply:
Doing `!subst("$funct3", funct3, "...")` leads to a cryptic tablegen error saying that the initializer could not be resolved. Maybe because `funct3` is an integer and not a string? Working around all of that might make it more complicated.